### PR TITLE
Add a new ``declare-non-slot`` error code

### PR DESCRIPTION
**Pull request by adamtuft**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Refs #9499 

Implement `declare-non-slot` which reports an error when a class has a `__slots__` member and a type hint on the class is not present in `__slots__`.

## Details

**Tests:**

- Add test case which should report the error code.
- Add test case where the error is suppressed due to `__dict__` in `__slots__`.

- Add test to detect annotation not in any base class's `__slots__`.

- Add test against false-positive when base doesn't have `__slots__` (since there is still then an instance `__dict__` thanks to base)

- Add test against false-positive when base has `__slots__` with `__dict__` entry.

- Ignore `declare-non-slots` for `regression_5479.py` which is for `assigning-non-slot`.

**Implementation:**

Implement check for `declare-non-slot` in `class_checker.py` and 2 new helper methods:

- Add method `_check_declare_non_slot` to report `declare-non-slot` error. This method checks `node` and all bases for a valid `__slots__`, gathering the names in all `__slots__` found. If `node` has an annotation not in any `__slots__`, then `declare-non-slot` is reported.

- Add helper method `_has_valid_slots` which returns `True` if a valid `__slots__` is found on a `ClassDef` (re-use logic from `_check_slots`).

- Refactor `_check_redefined_slots` to split out logic for getting `__slots__` names into `_get_classdef_slots_names` helper, since the same logic can be used by both functions.

**Some thoughts:**

- I thought this check belonged in the class checker instead of the dataclass checker since the problematic code is problematic even without the dataclass decorator. There may be other cases where the class annotations and __slots__ items should match, and making this check dataclass-specific would miss those cases.
- I went with `declare-non-slot` over `define-non-slot` as annotating a class feels more like a declaration that such a member exists, rather than a definition of that member.
- I made some changes to existing code to factor out some logic into a helper function `_get_classdef_slots_names` as this seemed the cleanest way to re-use existing logic. I hope this is ok.
- I've checked that the existing tests still pass and have run the primer tests with `python3 -m pytest -m primer_stdlib --primer-stdlib`.

**To Do:**

- Implement `missing-slot-annotation` from #9499
- ~Add documentation (good/bad code examples)~
- ~Add changelog/towncrier entry~

### CODE REVIEW

**Comment by adamtuft** [comment]
> 🤖 **Effect of this PR on checked open source code:** 🤖
> 
> 
> 
> **Effect on [pandas](https://github.com/pandas-dev/pandas):**
> The following messages are now emitted:
> 
> <details>
> 
> 1) declare-non-slot:
> *No such name 'values' in __slots__*
> https://github.com/pandas-dev/pandas/blob/bb0fcc23eed9f6a1a6506c6e27b98fb397ce747e/pandas/core/internals/blocks.py#L2097
> 2) declare-non-slot:
> *No such name 'values' in __slots__*
> https://github.com/pandas-dev/pandas/blob/bb0fcc23eed9f6a1a6506c6e27b98fb397ce747e/pandas/core/internals/blocks.py#L2147
> 
> </details>
> 
> 
> </details>
> 
> *This comment was generated for commit 245be58850744821a7de0c77afe7086a98ac7dc8*

Based on these messages, the error should probably exclude any aviation that looks like a descriptor.

**Review by Pierre-Sassoulas — COMMENTED** [review]
Looks great already, thank you. I don't have the time to really review the change at the moment but regarding the CI fail: You can create an example in the doc (see for example bad-chained-comparison : https://github.com/pylint-dev/pylint/tree/main/doc/data/messages/b/bad-chained-comparison) and a changelog entry for towncrier  (Some doc here, but you can copy paste an existing fragment of course: https://pylint.readthedocs.io/en/stable/development_guide/contributor_guide/contribute.html#creating-a-pull-request)

File: pylint/checkers/classes/class_checker.py
Comment (DanielNoord) [inline pylint/checkers/classes/class_checker.py:733]: Would you mind putting this under `E0244`? Just for consistency :)

File: pylint/checkers/classes/class_checker.py
Comment (DanielNoord) [inline pylint/checkers/classes/class_checker.py:734]:
```suggestion
        "No such name '%r' in __slots__",
```

File: pylint/checkers/classes/class_checker.py
Comment (DanielNoord) [inline pylint/checkers/classes/class_checker.py:896]: `_get_classdef_slots_names` also calls `_has_valid_slots`.

Can we just return early if `_get_classdef_slots_names` returns `[]`?

File: pylint/checkers/classes/class_checker.py
Comment (DanielNoord) [inline pylint/checkers/classes/class_checker.py:905]: @Pierre-Sassoulas Look at our fancy output :smile:

_Unrelated to this PR but this is the first time I see these annotations in the wild :)_

File: pylint/checkers/classes/class_checker.py
Comment (DanielNoord) [inline pylint/checkers/classes/class_checker.py:921]: could you add a confidence argument?

**Review by DanielNoord — COMMENTED** [review]
Left some comments. Code looks really good, thanks for opening the PR. I intended to have a look at your fork, but this is much easier as I can make in-line comments. Nice job on your first PR!

**Comment by adamtuft** [comment]
Thanks very much for the feedback! I'll incorporate suggestions when I have a little more time, hopefully in the next week or so.

File: pylint/checkers/classes/class_checker.py
Comment (adamtuft) [reply pylint/checkers/classes/class_checker.py:733]: Looks like `E0244` is already taken by `invalid-enum-extension`, this is the next available.

File: pylint/checkers/classes/class_checker.py
Comment (adamtuft) [reply pylint/checkers/classes/class_checker.py:921]: I'm not certain how to know what confidence level I should use. but went with `confidence=INFERENCE` as this seems the standard for the class checker (and I have used inference in the check).

File: pylint/checkers/classes/class_checker.py
Comment (Pierre-Sassoulas) [reply pylint/checkers/classes/class_checker.py]: I think this is going to create double simple quote ( "No such name ''c'' in __slots__") as the ``%r`` already call the repr function on a string.

File: pylint/checkers/classes/class_checker.py
Comment (Pierre-Sassoulas) [reply pylint/checkers/classes/class_checker.py:905]: I'm not sure I understood 😄 I don't see annotations here ?

File: pylint/checkers/classes/class_checker.py
Comment (Pierre-Sassoulas) [reply pylint/checkers/classes/class_checker.py:921]: Yes if you're using ``infer``, the confidence level is inference, otherwise HIGH. With the apparition of ruff, unless there's inference involved ruff can probably do it as well and faster than us.

File: pylint/checkers/classes/class_checker.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:737]:
```suggestion
        "Raised when a type annotation on a class is absent from the list of names in __slots__, "
        "and __slots__ does not contain a __dict__ entry.",
```

File: pylint/checkers/classes/class_checker.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:905]:
```suggestion
                for slot_name in self._get_classdef_slots_names(ancestor):
                    if slot_name == "__dict__":
                        return
                    slot_names.append(slot_name)
```
If I understood the code well, it seems we can cut all the following inferences if the a base contains ``__dict__`` which should be better performance wise as inference is costly. (based on intuition, not benchmark)

File: pylint/checkers/classes/class_checker.py
Comment (adamtuft) [reply pylint/checkers/classes/class_checker.py:896]: Good question... This got me thinking more about this implementation and I now think it has issues. For example, what to do when `__slots__` is present but empty? That's permitted if the class takes part in multiple inheritance. I also think it's unclear in my implementation whether I mean to just check for the presence of `__slots__` or to check that it is also a valid `__slots__` value. I'll need to think about this some more... :thinking:

File: pylint/checkers/classes/class_checker.py
Comment (adamtuft) [reply pylint/checkers/classes/class_checker.py:905]: Yes, this looks like a better idea to me, thanks!

**Comment by adamtuft** [comment]
Thanks so much for the review comments! I've reflected on my implementation and I think it needs a bit more work :thinking:  - just to share a few thoughts:

- Could we just return early if `_get_classdef_slots_names` returns `[]`?  https://github.com/pylint-dev/pylint/pull/9564#discussion_r1577417813
- What to do about empty `__slots__`, which is *required* if the class takes part in multiple inheritance [[link](https://docs.python.org/3/reference/datamodel.html#slots:~:text=Multiple%20inheritance%20with%20multiple%20slotted%20parent%20classes%20can%20be%20used%2C%20but%20only%20one%20parent%20is%20allowed%20to%20have%20attributes%20created%20by%20slots%20(the%20other%20bases%20must%20have%20empty%20slot%20layouts)%20%2D%20violations%20raise%20TypeError.)].
- Make it clearer under what conditions the check should return early, and whether to be permissive or conservative when trying to gather names from `__slots__` across base classes (I don't want to cause lots of false-positives!)

**Comment by DanielNoord** [comment]
Just letting you know that I do want to review this but haven't found the time. I hope to get to this in the next 7 days.

File: pylint/checkers/classes/class_checker.py
Comment (DanielNoord) [inline pylint/checkers/classes/class_checker.py:1539]: I'm on mobile but this seems very similar to `_check_slots`. I'm a bit worried that we're duplicating checks and run the risk of code drift.

Have you considered refactoring the other method to serve both purposes?

Comment (adamtuft) [reply pylint/checkers/classes/class_checker.py:1539]: I agree that `_check_slots` and `_has_valid_slots` are similar and that it would be best not to duplicate these checks. Refactoring is complicated by the fact that `_check_slots` is doing a few jobs at the same time:

- checking whether `__slots__` is valid
- if not, add 1 of 2 possible messages based on the reason why it isn't valid
- If `__slots__` looks valid, get the slot items and apply 2 further checks.

The difference between `_check_slots` and `_has_valid_slots` are that `_check_slots` cares about the reason why `__slots__` isn't valid, whereas `_has_valid_slots` only detects whether `__slots__` is valid. It might be wise to introduce a method that serves both purposes e.g. `_validate_slots` that returns an `Enum` that either reports a valid `__slots__` or says why it is invalid. This would allow you to separate the multiple concerns of the `_check_slots` method a little more clearly.

The question is whether refactoring `_check_slots` is within the scope of this MR. To me that feels like quite a big refactor that should be its own MR. I would propose that if the present MR is accepted an issue should then be opened to highlight the need to refactor these slot checks to reduce the duplication. I'd be happy to work on that issue as a separate MR.

**Review by DanielNoord — APPROVED** [review]
Thanks, I had another good look and think this is good to go.

@Pierre-Sassoulas do you want to do a second review? Or shall we just merge this? :)

File: pylint/checkers/classes/class_checker.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:925]: We have ``self.a = 42`` line 183 in the functional test, maybe getting coverage for this is as simple as adding a line with ``self.b : str = "AnnAssign.value is not None"`` too ?

**Review by Pierre-Sassoulas — COMMENTED** [review]
Just a nit if it's not too much work to fully cover the new code. This looks amazing, a great new check, thank you @adamtuft

File: pylint/checkers/classes/class_checker.py
Comment (adamtuft) [reply pylint/checkers/classes/class_checker.py:925]: Thanks, I've added this, hopefully this will fix the code coverage complaint. Waiting for the CI checks to run now.

**Comment by adamtuft** [comment]
Thank you both for the reviews again! I'm not sure why the codecov check is still failing: https://github.com/pylint-dev/pylint/pull/9564/checks?check_run_id=25859465941

File: pylint/checkers/classes/class_checker.py
Comment (Pierre-Sassoulas) [reply pylint/checkers/classes/class_checker.py:925]: There might be specific condition to enter this condition, (a "b" value in slot maybe ?)

Comment (adamtuft) [reply pylint/checkers/classes/class_checker.py:925]: The codecov test result seems to have changed now, it's claiming that a line isn't covered which must be covered by the tests which generate the error. I don't understand why the [codecov test](https://github.com/pylint-dev/pylint/pull/9564/checks?check_run_id=25859465941) is failing, are you able to shed any light on this please?

Comment (Pierre-Sassoulas) [reply pylint/checkers/classes/class_checker.py:923]: I tried something with ``b: str`` being declared. I assume the diff between main and this branch got too big. Also fixed the doc generation.

File: tests/functional/s/slots_checks.py
Comment (adamtuft) [reply tests/functional/s/slots_checks.py:171]: Thanks, I see now that I added that annotation to a test which doesn't report the message, so it wasn't covering the required line.

Yes, looks good, that doesn't raise the error :+1:  thank you for your help!

**Review by Pierre-Sassoulas — APPROVED** [review]
Great first contribution @adamtuft, congratulation on becoming a pylint contributor ! This will be released in 3.3.0 :)

**Comment by adamtuft** [comment]
Thank you very much @Pierre-Sassoulas and @DanielNoord for your help and feedback on this contribution! Very happy to have a contribution accepted :partying_face:

**Pull request merged**
